graphrep.pl

```diff
@@ -0,0 +1,114 @@
+% This file contains out internal representation for graphs.
+% our approach uses a graph-term from.
+% Design initially borrowed from prologsite.
+
+% :- dynamc graph/3.
+
+% A GRAPH is defined by graph(NAME,[NODE],[EDGE],[GRAPH]) where:
+% - NAME is a globally unique string.
+% - NODE is defined by node(NAME, [EDGE]) where:
+%   - NAME is a unique string to the graph.
+% - EDGE is is defined by dEdge(FROM,TO) where:
+%   - FROM/TO is the NAME of a NODE.
+
+% Graph validity constraints
+% - There must no be no edges to a node not found in [NODE].
+% - There must not be any duplicate edges in [EDGES].
+% - NAME in graph must be globally unique.
+
+% The example graph is [a]<->[b] [c]
+graph(ex1,
+    [
+        node(a,[dEdge(a,b)]),
+        node(b,[dEdge(b,a)]),
+        node(c,[])
+    ],
+    [
+        eEdge(a,b),
+        dEdge(b,a)
+    ], []).
+
+% A graph can have a subgraph. Suppose ex1 is a subgraph of ex2
+% This graph contains is [y]->[z]->[a]<->[b] [c]
+graph(ex2,
+    [
+        node(y,[dEdge(y,z)]),
+        node(z,[dEdge(z,a)]),
+        node(a,[dEdge(a,b)]),
+        node(b,[dEdge(b,a)]),
+        node(c,[])
+    ],
+    [
+        dEdge(y,z),
+        dEdge(z,a),
+        dEdge(a,b),
+        dEdge(b,a)
+    ],
+    [
+        graph(ex1,
+            [
+                node(a,[dEdge(a,b)]),
+                node(b,[dEdge(b,a)]),
+                node(c,[])
+            ],
+            [
+                eEdge(a,b),
+                dEdge(b,a)
+            ],[])
+    ]).
+
+% Graphs can have properties that are calculated by the following predicates:
+
+% Len(I,N): the number of edges in a graph.
+% - I is the graphID
+% - N is the number of edges
+len(I, N) :- graph(I, _, L2,_), len(L2, N).
+len([], N) :- N is 0.
+len([_|T], N) :-
+    len(T, N1),
+    N is N1+1.
+
+% dis(I, N1, N2, N): the distance from N1 to N2 in graph I.
+% - I is the graphID
+% - N1/N2 is From/To Node name
+% - N is the distance
+dis(I, N1, N2, N) :-
+    graph(I, _, EDGES, _),
+    path(N1, N2, EDGES, [], N).
+
+% path(I, N1, N2, PATH): the path from N1 to N2 in graph I.
+% This is broken idk why I can find the length but not the path.....
```

Comment:
Send help

Reply:
```prolog
% Public Api
% Graph is the graph being searched
% From is the origin
% To is the destination
% Path is the path to the node, if one exists
findPathOuter(Graph, From, To, Path) :-
    graph(Graph, Nodes, _, _),
    findPath(From, To, Nodes, [], Path).

findNode(Name, [node(Name,Edges)|Rest], node(Name,Edges)).
findNode(Name, [node(Other,_)|Rest], N) :-
    not(Name == Other),
    findNode(Name, Rest, N).

findPath(From, From, _, _, [From]).
findPath(From, To, Nodes, Visited, [From|Path]) :-
    \+ member(From, Visited),
    findNode(From, Nodes, node(From, Edges)),
    findPathForNode(Edges, To, Nodes, [From|Visited], Path).

findPathForNode([dEdge(Start,Next)|Rest], To, Nodes, Visited, Path) :-
    findPath(Next, To, Nodes, Visited, Path).
findPathForNode([dEdge(_,_)|Rest], To, Nodes, Visited, Path) :-
    findPathForNode(Rest, To, Nodes, Visited, Path).
```

Reply:
Probably could name stuff better, but you get the drift

Reply:
soooo many trampolines huh